### PR TITLE
Docs: Improve documentation about error return values

### PR DESCRIPTION
**Pull request by matusvalo**

### CODE REVIEW

File: docs/src/userguide/language_basics.rst [hunk 1]
```diff
@@ -687,19 +687,24 @@ as a contract with the caller. Here is an example:
...
 
 With this declaration, whenever an exception occurs inside ``spam``, it will
 immediately return with the value ``-1``.  From the caller's side, whenever
 a call to spam returns ``-1``, the caller will assume that an exception has
-occurred and can now process or propagate it.
+occurred and can now process or propagate it. Calling ``spam()`` is roughly equivalent to::
+
+    ret_val = spam()
+    if ret_val == -1:
+        raise ValueError()
```
Comment (da-woods) [inline docs/src/userguide/language_basics.rst:707]: I'm not sure this is quite right. It's closer to `if ret_val == -1: raise` (i.e. raise whatever error is currently existing).

I don't think this quite translates into Python pseudo-code though. This feels to me like having something a little wrong could cause more confusion than it solves.

I'd actually be happier with a simplified version of the generated C code:

```
ret_val = spam();
if (ret_val == -1) goto error_label;
```

but then this pre-supposes some knowledge of the Python C API, so may not be great.

In short I don't quite know what should go here.

File: docs/src/userguide/language_basics.rst [hunk 2]
```diff
@@ -732,13 +737,18 @@ form of exception value declaration
...
 
 In this case, Cython generates a call to :c:func:`PyErr_Occurred` if the exception value
 is returned, to make sure it really received an exception and not just a normal
-result.
+result. Calling ``spam()`` is roughly equivalent to::
+
+    ret_val = spam()
+    if ret_val == -1:
+        if PyErr_Occurred():
+            raise ValueError()
```
Comment (da-woods) [inline docs/src/userguide/language_basics.rst:751]: Same comment as above...

File: docs/src/userguide/language_basics.rst (hunk 1 above)
Comment (matusvalo) [reply docs/src/userguide/language_basics.rst:707]: > I'm not sure this is quite right. 

Yes these examples are simplified to express the high level logic. My ultimate goal was to do something similar to the python itertools documentation [1]. If there is consensus that this is wrong direction feel free to close this PR or just I can remove the controversial parts. I am also fine with converting it to the simplified C code. The examples are pretty simple so I think that even people not knowing C syntax will be able to extract the main point. (BTW. in my opinion any non-trivial usage of Cython requires at least basic understanding of C)


[1] https://docs.python.org/3/library/itertools.html#itertools.accumulate

Comment (matusvalo) [reply docs/src/userguide/language_basics.rst]: @da-woods I have migrated examples to C as you proposed.

Comment (matusvalo) [reply docs/src/userguide/language_basics.rst]: Question: maybe as a small improvement we can use `error_handler` instead of `error_label` to improve readability of example. What do you think?

Comment (da-woods) [reply docs/src/userguide/language_basics.rst]: I think C is probably a clearer way of showing it - with Python I was just getting hung up on the idea of an unhandled exception being there without a try/catch. It just didn't write make sense to me.

I think `error_handler` is better, yes.

Comment (matusvalo) [reply docs/src/userguide/language_basics.rst]: OK everything implemented.

File: docs/src/userguide/language_basics.rst [hunk 3]
```diff
@@ -824,6 +845,10 @@ Some things to note:
 
   .. note:: Pointers to functions are currently not supported by pure Python mode. (GitHub issue :issue:`4279`)
 
+* If the returning type of a ``cdef`` function with ``except *`` or ``@cython.exceptval(check=True)`` is C integer,
+  enum, float or pointer type, Cython is calling :c:func:`PyErr_Occurred` only when
```
Comment (da-woods) [inline docs/src/userguide/language_basics.rst:849]:
```suggestion
  enum, float or pointer type, Cython calls :c:func:`PyErr_Occurred` only when
```

File: docs/src/userguide/language_basics.rst [hunk 4]
```diff
@@ -750,18 +761,26 @@ There is also a third form of exception value declaration
...
                 ...
 
 This form causes Cython to generate a call to :c:func:`PyErr_Occurred` after
-*every* call to spam, regardless of what value it returns. If you have a
+*every* call to spam, regardless of what value it returns. Calling ``spam()`` is roughly translated to the following C code:
+
+.. code-block:: C
+
+    spam()
+    ret_val = spam();
```
Comment (da-woods) [inline docs/src/userguide/language_basics.rst:780]:
```suggestion
```

**Comment by da-woods** [comment]
Thanks @matusvalo
